src/java/test/org/apache/zookeeper/metrics/BaseTestMetricsProvider.java

```diff
@@ -0,0 +1,137 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.zookeeper.metrics;
+
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.zookeeper.metrics.impl.NullMetricsProvider;
+
+/**
+ * Simple MetricsProvider for tests.
+ */
+public abstract class BaseTestMetricsProvider implements MetricsProvider {
+
+    @Override
+    public void configure(Properties prprts) throws MetricsProviderLifeCycleException {
+    }
+
+    @Override
+    public void start() throws MetricsProviderLifeCycleException {
+    }
+
+    @Override
+    public MetricsContext getRootContext() {
+        return NullMetricsProvider.NullMetricsContext.INSTANCE;
+    }
+
+    @Override
+    public void stop() {
+    }
+
+    public static final class MetricsProviderCapturingLifecycle extends BaseTestMetricsProvider {
+
+        public static final AtomicBoolean configureCalled = new AtomicBoolean();
```

Comment:
This kind of stuff (tracking calls) is exactly the thing that can be nicely achieved with Mockito mocks and/or spies in unit tests.

https://www.baeldung.com/mockito-verify

Would you please consider replace these test classes with mocks?

Reply:
@anmolnar 

The problem here is that the object is instantiated using the classname, so I cannot inject a mock object. And injecting objects into the runtime in this case will make the test mostly useless, as I want to test exactly the lifecycle of the object.

I could 'PowerMock#replace' many methods of NullMetricsProvider, but IMHO it is adding some extra complexity which is not worth.

Reply:
I see, that makes sense.